<div id="readme" class="Box-body readme blob instapaper_body js-code-block-container">
    <article class="markdown-body entry-content p-3 p-md-6" itemprop="text"><h1><a id="user-content-capítulo-15-classes-e-objetos" class="anchor" aria-hidden="true" href="#capítulo-15-classes-e-objetos"><svg class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Capítulo 15: Classes e objetos</h1>
<p>A esta altura você já sabe como usar funções para organizar código e tipos integrados para organizar dados. O próximo passo é aprender “programação orientada a objeto”, que usa tipos definidos pelos programadores para organizar tanto o código quanto os dados. A programação orientada a objeto é um tópico abrangente; será preciso passar por alguns capítulos para abordar o tema.</p>
<p>Os exemplos de código deste capítulo estão disponíveis em <a href="http://thinkpython2.com/code/Point1.py" rel="nofollow">http://thinkpython2.com/code/Point1.py</a>; as soluções para os exercícios estão disponíveis em <a href="http://thinkpython2.com/code/Point1_soln.py" rel="nofollow">http://thinkpython2.com/code/Point1_soln.py</a>.</p>
<h2><a id="user-content-151---tipos-definidos-pelos-programadores" class="anchor" aria-hidden="true" href="#151---tipos-definidos-pelos-programadores"><svg class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>15.1 - Tipos definidos pelos programadores</h2>
<p>Já usamos muitos tipos integrados do Python; agora vamos definir um tipo próprio. Como exemplo, criaremos um tipo chamado <code>Point</code>, que representa um ponto no espaço bidimensional.</p>
<p>Na notação matemática, os pontos muitas vezes são escritos entre parênteses, com uma vírgula separando as coordenadas. Por exemplo, (0,0) representa a origem e (x, y) representa o ponto que está x unidades à direita e y unidades acima da origem.</p>
<p>Há várias formas para representar pontos no Python:</p>
<ul>
<li>
<p>Podemos armazenar as coordenadas separadamente em duas variáveis, x e y.</p>
</li>
<li>
<p>Podemos armazenar as coordenadas como elementos em uma lista ou tupla.</p>
</li>
<li>
<p>Podemos criar um tipo para representar pontos como objetos.</p>
</li>
</ul>
<p>Criar um tipo é mais complicado que outras opções, mas tem vantagens que logo ficarão evidentes.</p>
<p>Um tipo definido pelo programador também é chamado de <b>classe</b>. Uma definição de classe pode ser assim:</p>

In [31]:
class Point:
    """Represents a point in 2-D space."""

<p>O cabeçalho indica que a nova classe se chama <code>Point</code>. O corpo é uma docstring que explica para que a classe serve. Você pode definir variáveis e métodos dentro de uma definição de classe, mas voltaremos a isso depois.</p>
<p>Definir uma classe denominada <code>Point</code> cria um <b>objeto de classe</b>:</p>

In [32]:
>>> Point

__main__.Point

<p>Como <code>Point</code> é definido no nível superior, seu “nome completo” é <code>__main__.Point</code>.</p>
<p>O objeto de classe é como uma fábrica para criar objetos. Para criar um <code>Point</code>, você chama <code>Point</code> como se fosse uma função:</p>

In [33]:
>>> blank = Point()
>>> blank

<p>O valor de retorno é uma referência a um objeto <code>Point</code>, que atribuímos a <code>blank</code>.</p>
<p>Criar um objeto chama-se <b>instanciação</b>, e o objeto é uma <b>instância</b> da classe.</p>
<p>Quando você exibe uma instância, o Python diz a que classe ela pertence e onde está armazenada na memória (o prefixo 0x significa que o número seguinte está em formato hexadecimal).</p>
<p>Cada objeto é uma instância de alguma classe, então “objeto” e “instância” são intercambiáveis. Porém, neste capítulo uso “instância” para indicar que estou falando sobre um tipo definido pelo programador.</p>
<h2><a id="user-content-152---atributos" class="anchor" aria-hidden="true" href="#152---atributos"><svg class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>15.2 - Atributos</h2>
<p>Você pode atribuir valores a uma instância usando a notação de ponto:</p>

In [34]:
>>> blank.x = 3.0
>>> blank.y = 4.0

<p>Essa sintaxe é semelhante à usada para selecionar uma variável de um módulo, como <code>math.pi</code> ou <code>string.whitespace</code>. Nesse caso, entretanto, estamos atribuindo valores a elementos nomeados de um objeto. Esses elementos chamam-se <b>atributos</b>.</p>
<p>Em inglês, quando é um substantivo, a palavra “AT-trib-ute” é pronunciada com ênfase na primeira sílaba, ao contrário de “a-TRIB-ute”, que é um verbo.</p>
<p>O diagrama seguinte mostra o resultado dessas atribuições. Um diagrama de estado que mostra um objeto e seus atributos chama-se <b>diagrama de objeto</b>; veja a Figura 15.1.</p>
<p><a target="_blank" rel="noopener noreferrer" href="https://github.com/PenseAllen/PensePython2e/raw/master/fig/tnkp_1501.png"><img src="https://github.com/PenseAllen/PensePython2e/raw/master/fig/tnkp_1501.png" alt="Figura 15.1 – Diagrama de um objeto Point." style="max-width:100%;"></a>
<br><em>Figura 15.1 – Diagrama de um objeto</em> <code>Point</code>.</p>
<p>A variável <code>blank</code> refere-se a um objeto <code>Point</code>, que contém dois atributos. Cada atributo refere-se a um número de ponto flutuante.</p>
<p>Você pode ler o valor de um atributo usando a mesma sintaxe:</p>

In [35]:
>>> blank.y

4.0

In [36]:
>>> x = blank.x
>>> x

3.0

<p>A expressão <code>blank.x</code> significa “Vá ao objeto a que <code>blank</code> se refere e pegue o valor de <code>x</code>”. No exemplo, atribuímos este valor a uma variável <code>x</code>. Não há nenhum conflito entre a variável <code>x</code> e o atributo <code>x</code>.</p>
<p>Você pode usar a notação de ponto como parte de qualquer expressão. Por exemplo:</p>

In [37]:
>>> '(%g, %g)' % (blank.x, blank.y)

'(3, 4)'

In [38]:
>>> import math
>>> distance = math.sqrt(blank.x ** 2 + blank.y **2)
>>> distance

5.0

<p>Você pode passar uma instância como argumento da forma habitual. Por exemplo:</p>

In [39]:
def print_point(p):
    print('(%g, %g)' % (p.x, p.y))

<p><code>print_point</code> toma um ponto como argumento e o exibe em notação matemática. Para invocá-lo, você pode passar <code>blank</code> como argumento:</p>

In [40]:
>>> print_point(blank)

(3, 4)


<p>Dentro da função, <code>p</code> é um <i>alias</i> para <code>blank</code>, então, se a função altera <code>p</code>, <code>blank</code> também muda.</p>
<p>Como exercício, escreva uma função chamada <code>distance_between_points</code>, que toma dois pontos como argumentos e retorna a distância entre eles.</p>
<h2><a id="user-content-153---retângulos" class="anchor" aria-hidden="true" href="#153---retângulos"><svg class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>15.3 - Retângulos</h2>
<p>Às vezes, é óbvio quais deveriam ser os atributos de um objeto, mas outras é preciso decidir entre as possibilidades. Por exemplo, vamos supor que você esteja criando uma classe para representar retângulos. Que atributos usaria para especificar a posição e o tamanho de um retângulo? Você pode ignorar ângulo; para manter as coisas simples, suponha que o retângulo seja vertical ou horizontal.</p>
<p>Há duas possibilidades, no mínimo:</p>
<ul>
<li>
<p>Você pode especificar um canto do retângulo (ou o centro), a largura e a altura.</p>
</li>
<li>
<p>Você pode especificar dois cantos opostos.</p>
</li>
</ul>
<p>Nesse ponto é difícil dizer qual opção é melhor, então implementaremos a primeira, como exemplo.</p>
<p>Aqui está a definição de classe:</p>

In [42]:
class Rectangle:
    """Represents a rectangle.
    
    attributes: width, height, corner.
    """

<p>A <i>docstring</i> lista os atributos: <code>width</code> e <code>height</code> são números; <code>corner</code> é um objeto <code>Point</code> que especifica o canto inferior esquerdo.</p>
<p>Para representar um retângulo, você tem que instanciar um objeto <code>Rectangle</code> e atribuir valores aos atributos:</p>

In [43]:
box = Rectangle()
box.width = 100.0
box.height = 200.0
box.corner = Point()
box.corner.x = 0.0
box.corner.y = 0.0

<p>A expressão <code>box.corner.x</code> significa “Vá ao objeto ao qual <code>box</code> se refere e pegue o atributo denominado <code>corner</code>; então vá a este objeto e pegue o atributo denominado <code>x</code>”.</p>
<p>A Figura 15.2 mostra o estado deste objeto. Um objeto que é um atributo de outro objeto é <b>integrado</b>.</p>
<p><a target="_blank" rel="noopener noreferrer" href="https://github.com/PenseAllen/PensePython2e/raw/master/fig/tnkp_1502.png"><img src="https://github.com/PenseAllen/PensePython2e/raw/master/fig/tnkp_1502.png" alt="Figura 15.2 – Diagrama de um objeto Rectangle." style="max-width:100%;"></a>
<br><em>Figura 15.2 – Diagrama de um objeto</em> <code>Rectangle</code>.</p>
<h2><a id="user-content-154---instâncias-como-valores-de-retorno" class="anchor" aria-hidden="true" href="#154---instâncias-como-valores-de-retorno"><svg class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>15.4 - Instâncias como valores de retorno</h2>
<p>As funções podem retornar instâncias. Por exemplo, <code>find_center</code> recebe um <code>Rectangle</code> como argumento e devolve um <code>Point</code>, que contém as coordenadas do centro do retângulo:</p>

In [45]:
def find_center(rect):
    p = Point()
    p.x = rect.corner.x + rect.width/2
    p.y = rect.corner.y + rect.height/2
    return p

<p>Aqui está um exemplo que passa <code>box</code> como um argumento para <code>find_center</code> e atribui o <code>Point</code> resultante à variável <code>center</code>:</p>

In [46]:
>>> center = find_center(box)
>>> print_point(center)

(50, 100)


<h2><a id="user-content-155---objetos-são-mutáveis" class="anchor" aria-hidden="true" href="#155---objetos-são-mutáveis"><svg class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>15.5 - Objetos são mutáveis</h2>
<p>Você pode alterar o estado de um objeto fazendo uma atribuição a um dos seus atributos. Por exemplo, para mudar o tamanho de um retângulo sem mudar sua posição, você pode alterar os valores de <code>width</code> e <code>height</code>:</p>

In [47]:
box.width = box.width + 50
box.height = box.height + 100

<p>Você também pode escrever funções que alteram objetos. Por exemplo, <code>grow_rectangle</code> recebe um objeto <code>Rectangle</code> e dois números, <code>dwidth</code> e <code>dheight</code>, e adiciona os números à largura e altura do retângulo:</p>

In [48]:
def grow_rectangle(rect, dwidth, dheight):
    rect.width += dwidth
    rect.height += dheight

<p>Eis um exemplo que demonstra o efeito:</p>

In [49]:
>>> box.width, box.height

(150.0, 300.0)

In [50]:
>>> grow_rectangle(box, 50, 100)
>>> box.width, box.height

(200.0, 400.0)

<p>Dentro da função, <code>rect</code> é um <i>alias</i> de <code>box</code>, então quando a função altera <code>rect</code>, <code>box</code> aponta para o objeto alterado.</p>
<p>Como exercício, escreva uma função chamada <code>move_rectangle</code> que toma um Rectangle e dois números chamados <code>dx</code> e <code>dy</code>. Ela deve alterar a posição do retângulo, adicionando <code>dx</code> à coordenada <code>x</code> de <code>corner</code> e adicionando <code>dy</code> à coordenada <code>y</code> de <code>corner</code>.</p>
<h2><a id="user-content-156---cópia" class="anchor" aria-hidden="true" href="#156---cópia"><svg class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>15.6 - Cópia</h2>
<p><i>Alias</i> podem tornar um programa difícil de ler porque as alterações em um lugar podem ter efeitos inesperados em outro lugar. É difícil monitorar todas as variáveis que podem referir-se a um dado objeto.</p>
<p>Em vez de usar <i>alias</i>, copiar o objeto pode ser uma alternativa. O módulo <code>copy</code> contém uma função chamada <code>copy</code> que pode duplicar qualquer objeto:</p>

In [51]:
>>> p1 = Point()
>>> p1.x = 3.0
>>> p1.y = 4.0
>>> import copy
>>> p2 = copy.copy(p1)

<p><code>p1</code> e <code>p2</code> contêm os mesmos dados, mas não são o mesmo <code>Point</code>:</p>

In [52]:
>>> print_point(p1)

(3, 4)


In [53]:
>>> print_point(p2)

(3, 4)


In [54]:
>>> p1 is p2

False

In [55]:
>>> p1 == p2

False

<p>O operador <code>is</code> indica que <code>p1</code> e <code>p2</code> não são o mesmo objeto, que é o que esperamos. Porém, você poderia ter esperado que <code>==</code> fosse apresentado como <code>True</code>, porque esses pontos contêm os mesmos dados. Nesse caso, pode ficar desapontado ao saber que, para instâncias, o comportamento padrão do operador <code>==</code> é o mesmo que o do operador <code>is</code>; ele verifica a identidade dos objetos, não a sua equivalência. Isso acontece porque, para tipos definidos pelo programador, o Python não sabe o que deve ser considerado equivalente. Pelo menos, ainda não.</p>
<p>Se você usar <code>copy.copy</code> para duplicar um retângulo, descobrirá que ele copia o objeto <code>Rectangle</code>, mas não o <code>Point</code> embutido nele:</p>

In [56]:
>>> box2 = copy.copy(box)
>>> box2 is box

False

In [57]:
>>> box2.corner is box.corner

True

<p>A Figura 15.3 mostra como fica o diagrama de objeto. Esta operação chama-se <b>cópia superficial</b> porque copia o objeto e qualquer referência que contenha, mas não os objetos integrados.</p>
<p><a target="_blank" rel="noopener noreferrer" href="https://github.com/PenseAllen/PensePython2e/raw/master/fig/tnkp_1503.png"><img src="https://github.com/PenseAllen/PensePython2e/raw/master/fig/tnkp_1503.png" alt="Figura 15.3 – Diagrama: dois objetos Rectangle compartilhando o mesmo Point." style="max-width:100%;"></a>
<br><em>Figura 15.3 – Diagrama: dois objetos</em> <code>Rectangle</code> <em>compartilhando o mesmo</em> <code>Point</code>.</p>
<p>Para a maior parte das aplicações, não é isso que você quer. Nesse exemplo, invocar <code>grow_rectangle</code> em um dos <code>Rectangle</code>s não afetaria o outro, mas invocar <code>move_rectangle</code> em qualquer um deles afetaria a ambos! Esse comportamento é confuso e propenso a erros.</p>
<p>Felizmente, o módulo <code>copy</code> oferece um método chamado <code>deepcopy</code> que copia não só o objeto, mas também os objetos aos quais ele se refere, e os objetos aos quais estes se referem, e assim por diante. Você não se surpreenderá ao descobrir que esta operação se chama <b>cópia profunda</b>.</p>

In [58]:
>>> box3 = copy.deepcopy(box)
>>> box3 is box

False

In [59]:
>>> box3.corner is box.corner

False

<code>box3</code> e <code>box</code> são objetos completamente separados.</pre></div>
<p>Como exercício, escreva uma versão de <code>move_rectangle</code> que cria e retorne um novo retângulo em vez de alterar o antigo.</p>
<h2><a id="user-content-157---depuração" class="anchor" aria-hidden="true" href="#157---depuração"><svg class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>15.7 - Depuração</h2>
<p>Ao começar a trabalhar com objetos, provavelmente você encontrará algumas novas exceções. Se tentar acessar um atributo que não existe, recebe um <code>AttributeError</code>:</p>

In [60]:
>>> p = Point()
>>> p.x = 3
>>> p.y = 4
>>> p.z

AttributeError: 'Point' object has no attribute 'z'

<p>Se não estiver certo sobre o tipo que um objeto é, pode perguntar:</p>

In [61]:
>>> type(p)

__main__.Point

<p>Você também pode usar <code>isinstance</code> para verificar se um objeto é uma instância de uma classe:</p>

In [62]:
>>> isinstance(p, Point)

True

<p>Caso não tenha certeza se um objeto tem determinado atributo, você pode usar a função integrada <code>hasattr</code>:</p>

In [63]:
>>> hasattr(p, 'x')

True

In [64]:
>>> hasattr(p, 'z')

False

<p>O primeiro argumento pode ser qualquer objeto; o segundo argumento é uma <code>string</code> com o nome do atributo.</p>
<p>Você também pode usar uma instrução <code>try</code> para ver se o objeto tem os atributos de que precisa:</p>

In [66]:
try:
    x = p.x
except AttributeError:
    x = 0

<p>Essa abordagem pode facilitar a escrita de funções que atuam com tipos diferentes; você verá mais informações sobre isso em “Polimorfismo”, no Capítulo 17.</p>
<h2><a id="user-content-158---glossário" class="anchor" aria-hidden="true" href="#158---glossário"><svg class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>15.8 - Glossário</h2>
<dl>
<dt><a id="user-content-glos:classe"><b>classe</b></a></dt>
<dd>Tipo definido pelo programador. Uma definição de classe cria um objeto de classe.</dd>
<dt><a id="user-content-glos:objeto de classe"><b>objeto de classe</b></a></dt>
<dd>Objeto que contém a informação sobre um tipo definido pelo programador. O objeto de classe pode ser usado para criar instâncias do tipo.</dd>
<dt><a id="user-content-glos:instância"><b>instância</b></a></dt>
<dd>Objeto que pertence a uma classe.</dd>
<dt><a id="user-content-glos:instanciar"><b>instanciar</b></a></dt>
<dd>Criar um objeto.</dd>
<dt><a id="user-content-glos:atributo"><b>atributo</b></a></dt>
<dd>Um dos valores denominados associados a um objeto.</dd>
<dt><a id="user-content-glos:objeto integrado"><b>objeto integrado</b></a></dt>
<dd>Objeto que é armazenado como um atributo de outro objeto.</dd>
<dt><a id="user-content-glos:cópia superficial"><b>cópia superficial</b></a></dt>
<dd>Copiar o conteúdo de um objeto, inclusive qualquer referência a objetos integrados; implementada pela função copy no módulo copy.</dd>
<dt><a id="user-content-glos:cópia profunda"><b>cópia profunda</b></a></dt>
<dd>Copiar o conteúdo de um objeto, bem como qualquer objeto integrado, e qualquer objeto integrado a estes, e assim por diante; implementado pela função deepcopy no módulo copy.</dd>
<dt><a id="user-content-glos:diagrama de objeto"><b>diagrama de objeto</b></a></dt>
<dd>Diagrama que mostra objetos, seus atributos e os valores dos atributos.</dd>
</dl>
<h2><a id="user-content-159---exercícios" class="anchor" aria-hidden="true" href="#159---exercícios"><svg class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>15.9 - Exercícios</h2>
<h3><a id="user-content-exercício-151" class="anchor" aria-hidden="true" href="#exercício-151"><svg class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Exercício 15.1</h3>
<ol>
<li>
<p>Escreva uma definição para uma classe denominada <code>Circle</code>, com os atributos <code>center</code> e <code>radius</code>, onde <code>center</code> é um objeto <code>Point</code> e <code>radius</code> é um número.</p>
</li>
<li>
<p>Instancie um objeto <code>Circle</code>, que represente um círculo com o centro em 150, 100 e raio 75.</p>
</li>
<li>
<p>Escreva uma função denominada <code>point_in_circle</code>, que tome um <code>Circle</code> e um <code>Point</code> e retorne <code>True</code>, se o ponto estiver dentro ou no limite do círculo.</p>
</li>
<li>
<p>Escreva uma função chamada <code>rect_in_circle</code>, que tome um <code>Circle</code> e um <code>Rectangle</code> e retorne <code>True</code>, se o retângulo estiver totalmente dentro ou no limite do círculo.</p>
</li>
<li>
<p>Escreva uma função denominada <code>rect_circle_overlap</code>, que tome um <code>Circle</code> e um <code>Rectangle</code> e retorne <code>True</code>, se algum dos cantos do retângulo cair dentro do círculo. Ou, em uma versão mais desafiadora, retorne <code>True</code> se alguma parte do retângulo cair dentro do círculo.</p>
</li>
</ol>
<p>Solução: <a href="http://thinkpython2.com/code/Circle.py" rel="nofollow">http://thinkpython2.com/code/Circle.py</a>.</p>
<h3><a id="user-content-exercício-152" class="anchor" aria-hidden="true" href="#exercício-152"><svg class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg></a>Exercício 15.2</h3>
<ol>
<li>
<p>Escreva uma função chamada <code>draw_rect</code> que receba um objeto <code>Turtle</code> e um <code>Rectangle</code> e use o <code>Turtle</code> para desenhar o retângulo. Veja no Capítulo 4 os exemplos de uso de objetos <code>Turtle</code>.</p>
</li>
<li>
<p>Escreva uma função chamada <code>draw_circle</code>, que tome um <code>Turtle</code> e um <code>Circle</code> e desenhe o círculo.</p>
</li>
</ol>